### PR TITLE
Implement check checking

### PR DESCRIPTION
**Pull request by alxmjo**
- Refactor movePiece into two functions, one that checks the condition
(isValidMove()) and one that moves the piece (movePiece()). This allows
us to use the condition check for isInCheck().
- Add several other functions to board which do what they say, namely
getPieces(), getPieceLocations(), and getKingLocation().
- Implement isInCheck() at the Game level (since it's more a game state
then a board state).
- Add some organizational comments to Piece.hpp.
- Add unit tests for Board functions.
- Create a unit test file for Game and include a unit test which tests
whether or not different colors are in check.

### CODE REVIEW

File: source/Board.cpp [hunk 1]
```diff
@@ -554,5 +535,86 @@ bool Board::isKnightMove(const std::pair<int, int>& fromCoords, const std::pair<
...
+	// Check if from location is occupied by piece of same color
+	if (isOccupiedSameColor(fromCoords, toCoords))
+	{
+		std::cout << "Error: Cannot move from " << intToAlgebraic(fromCoords) << " to " << intToAlgebraic(toCoords)
+				  << " because there's a friendly piece at " << intToAlgebraic(toCoords) << '\n';
+		return false;
+	}
+
+	// Check if this is a valid move for this piece
+	const Piece *piece = getPiece(fromCoords);
```
Comment (djbauman) [inline source/Board.cpp]:
```suggestion
	const Piece *piece = getPiece(fromCoords);
	if (piece == nullptr)
	{
		std::cout << "Error: No piece at this starting square" << '\n';
		return false;
	}
```

Comment (djbauman) [reply source/Board.cpp]: I noticed that I can crash the game by inputting something like "e6 e8" when there's no piece at e6. Easy to fix with a small change like this.

File: source/Board.cpp [hunk 2]
```diff
@@ -554,5 +535,86 @@ bool Board::isKnightMove(const std::pair<int, int>& fromCoords, const std::pair<
...
+	std::cout << "Attempting to move from " << intToAlgebraic(fromCoords) << " to " << intToAlgebraic(toCoords) << '\n';
+
+	// Check if from and to locations are on the board
+	if (!isOnBoard(fromCoords) || !isOnBoard(toCoords))
+	{
+		std::cout << "Error: Coords are out of bounds" << '\n';
+		return false;
+	}
+
+	// Check if from location is occupied by piece of same color
```
Comment (djbauman) [inline source/Board.cpp]:
```suggestion
	// Check if to location is occupied by piece of same color
```

**Comment by alxmjo** [comment]
Thanks for the extra-close read on this, Daniel!

**Comment by alxmjo** [comment]
At some point I'd like to talk/think through which status messages we want to print and how. For example, the messages in `isValidMove()` get printed a ton since that function is used repeatedly for checking whether or not a player is in check.

**Pull request merged**